src/nvim/profile.c

```diff
@@ -944,3 +944,50 @@ void time_msg(const char *mesg, const proftime_T *start)
   g_prev_time = now;
   fprintf(time_fd, ": %s\n", mesg);
 }
+
+/// Initializes the time time_fd stream used to write startup times
+///
+/// @param startup_time_file the startuptime report file path
+/// @param process_name the name of the current process to write in the report.
+void time_init(const char *startup_time_file, const char *process_name)
+{
+  time_fd = fopen(startup_time_file, "a");
+  if (time_fd == NULL) {
+    semsg(_(e_notopen), startup_time_file);
+    return;
+  }
+  startuptime_buf = xmalloc(sizeof(char) * (STARTUP_TIME_BUF_SIZE + 1));
+  // the startuptime file is (potentially) bein written by several nvim processes concurrently.
+  // in order to avoid messing up the file with concurrent writes all startuptimes are written
+  // to a memory buffer first and flushed to disk only after the process is done with the
+  // is done with the initialization.
+  // in order to achieve this we set a large-enough buffer to the file stream, big enough to store
+  // all startup times. The _IOFBF mode makes sure that the buffer is not automatically flushed
+  // ("controlled buffering").
+  // The times are flushed to disk manually when "time_finish" is called.
+  if (setvbuf(time_fd, startuptime_buf, _IOFBF, STARTUP_TIME_BUF_SIZE + 1) != 0) {
+    xfree(startuptime_buf);
+    fclose(time_fd);
+    time_fd = NULL;
+    semsg("Error initializing startup time");
+    return;
```

Comment:
untested but should work:

```suggestion
  int r = setvbuf(time_fd, startuptime_buf, _IOFBF, STARTUP_TIME_BUF_SIZE + 1);
  if (r != 0) {
    xfree(startuptime_buf);
    fclose(time_fd);
    time_fd = NULL;
    // Might as well ELOG also I guess.
    ELOG("time_init: setvbuf failed: %d %s", r, uv_err_name(r));
    semsg("time_init: setvbuf failed: %d %s", r, uv_err_name(r));
    return;
```